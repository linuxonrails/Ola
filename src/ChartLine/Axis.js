import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const ChartLineAxis = ({ max, steps, className, format }) => {
  const styles = cx('ola_chartLine-axis', className)

  steps = Math.min(max, steps)
  const decimals = max < steps * steps ? 10 : 1
  const step = max > 0 ? max / steps : 0
  const children = new Array(steps + 1).fill(null)
    .map((v, index) => <span key={index}>{format(Math.round(index * step * decimals) / decimals)}</span>)

  return (
    <div className={styles}>
      { children }
    </div>
  )
}

ChartLineAxis.defaultProps = {
  steps: 4,
  max: null,
  className: null,
  format:(num)=> num
}

ChartLineAxis.propTypes = {
  /** Max value */
  max: PT.number.isRequired,
  /** Number of steps */
  steps: PT.number,
  /** Extra className */
  className: PT.string,
  /** Function to return number */
  format:PT.func
}

export default ChartLineAxis
