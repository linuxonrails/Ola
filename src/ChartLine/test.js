import React from 'react'
import ChartLine from './'
import ChartLineLabel from './Label'
import ChartLineAxis from './Axis'
import { getRanges } from '../utils'
import renderer from 'react-test-renderer'

const mockData = [
  {
    value: 5,
    label: '5 keywords',
  },
  {
    value: 2,
    label: '2 keywords',
  },
  {
    value: 1,
    label: '1 keywords',
  },
  {
    value: 10,
    label: '10 keywords',
  },
]

it('Default ChartLine', () => {
  const tree = renderer.create(<ChartLine>Test</ChartLine>).toJSON()
  expect(tree).toMatchSnapshot()
})

it('Default ChartLineLabel', () => {
  const tree = renderer.create(<ChartLineLabel></ChartLineLabel>).toJSON()
  expect(tree).toMatchSnapshot()
})

it('ChartLine and ChartLineLabel', () => {
  const ranges = getRanges(mockData.map((data) => data.value))
  const tree = renderer
    .create(
      <ChartLine ranges={ranges}>
        {mockData.map((row, idx) => (
          <ChartLineLabel key={idx} value={ranges[idx]}>
            {row.label}
          </ChartLineLabel>
        ))}
      </ChartLine>,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ChartLineAxis with max value zero', () => {
  const mockDataNull = [
    {
      value: null,
    },
    {
      value: null,
    },
    {
      value: null,
    },
    {
      value: 0,
    },
  ]
  const ranges = getRanges(mockDataNull.map((data) => data.value))
  const links = [null, null, null, 0]
  const maxValue = Math.max(...links)
  const tree = renderer
    .create(
      <ChartLine ranges={ranges}>
        <ChartLineAxis max={maxValue} />
        {mockDataNull.map((row, idx) => (
          <ChartLineLabel key={idx} value={ranges[idx]} />
        ))}
      </ChartLine>,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
