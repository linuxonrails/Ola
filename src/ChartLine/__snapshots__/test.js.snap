// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLine and ChartLineLabel 1`] = `
<div
  className="ola_chartLine"
>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 0.5,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    >
      5 keywords
    </div>
    <span
      className="ola_chartLine-point"
      style={
        Object {
          "--color": undefined,
          "--value": 0.5,
        }
      }
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 0.2,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    >
      2 keywords
    </div>
    <span
      className="ola_chartLine-point"
      style={
        Object {
          "--color": undefined,
          "--value": 0.2,
        }
      }
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 0.1,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    >
      1 keywords
    </div>
    <span
      className="ola_chartLine-point"
      style={
        Object {
          "--color": undefined,
          "--value": 0.1,
        }
      }
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 1,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    >
      10 keywords
    </div>
    <span
      className="ola_chartLine-point"
      style={
        Object {
          "--color": undefined,
          "--value": 1,
        }
      }
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <svg
    className="ola_chartLine-svg"
    preserveAspectRatio="none"
    role="img"
    viewBox="0 0 100 160"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      className="ola_chartLine-svg-background"
      d="M 0 0 M 0 80 L 12.5 80 L 37.5 125 L 62.5 140 L 87.5 5 L 100 5 V 0 H 0 Z"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
    <path
      className="ola_chartLine-svg-line"
      d="M 0 80 L 12.5 80 L 37.5 125 L 62.5 140 L 87.5 5 L 100 5"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
  </svg>
</div>
`;

exports[`ChartLineAxis with max value zero 1`] = `
<div
  className="ola_chartLine"
>
  <div
    className="ola_chartLine-axis"
  >
    <span>
      0
    </span>
  </div>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 0,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 0,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 0,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <div
    className="ola_chartLine-value"
    style={
      Object {
        "--value": 0,
      }
    }
  >
    <div
      className="ols_chartLine-label"
    />
    <span
      className="ola_chartLine-point"
      style={
        Object {
          "--color": undefined,
          "--value": 0,
        }
      }
    />
    <div
      className="ola_chartLine-space"
    />
    <div
      className="ola_chartLine-footer"
    />
  </div>
  <svg
    className="ola_chartLine-svg"
    preserveAspectRatio="none"
    role="img"
    viewBox="0 0 100 160"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      className="ola_chartLine-svg-background"
      d="M 0 0 M 0 155 L 12.5 155 L 37.5 155 L 62.5 155 L 87.5 155 L 100 155 V 0 H 0 Z"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
    <path
      className="ola_chartLine-svg-line"
      d="M 87.5 155 L 87.5 155 L 100 155"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
  </svg>
</div>
`;

exports[`Default ChartLine 1`] = `
<div
  className="ola_chartLine"
>
  Test
  <svg
    className="ola_chartLine-svg"
    preserveAspectRatio="none"
    role="img"
    viewBox="0 0 100 160"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      className="ola_chartLine-svg-background"
      d="M 0 0  V 0 H 0 Z"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
  </svg>
</div>
`;

exports[`Default ChartLineLabel 1`] = `
<div
  className="ola_chartLine-value"
  style={
    Object {
      "--value": 0,
    }
  }
>
  <div
    className="ols_chartLine-label"
  />
  <span
    className="ola_chartLine-point"
    style={
      Object {
        "--color": undefined,
        "--value": 0,
      }
    }
  />
  <div
    className="ola_chartLine-space"
  />
  <div
    className="ola_chartLine-footer"
  />
</div>
`;
